README.md
# Phaser3 Nine Patch

[![GitHub license](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/koreezgames/phaser3-ninepatch-plugin/blob/master/LICENSE)
[![Build Status](https://secure.travis-ci.org/koreezgames/phaser3-ninepatch-plugin.svg?branch=master)](https://travis-ci.org/koreezgames/phaser3-ninepatch-plugin)
[![codebeat badge](https://codebeat.co/badges/e3792494-1875-4826-be00-2124148b9287)](https://codebeat.co/projects/github-com-koreezgames-phaser3-ninepatch-plugin-master)
[![Test Coverage](https://api.codeclimate.com/v1/badges/63e0c72189fa97ca55db/test_coverage)](https://codeclimate.com/github/koreezgames/phaser3-ninepatch-plugin/test_coverage)
[![npm version](https://badge.fury.io/js/%40koreez%2Fphaser3-ninepatch.svg)](https://badge.fury.io/js/%40koreez%2Fphaser3-ninepatch)
[![Greenkeeper badge](https://badges.greenkeeper.io/koreezgames/phaser3-ninepatch-plugin.svg)](https://greenkeeper.io/)
[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)

Phaser3 Nine Patch plugin adds 9-slice scaling support to Phaser3!

#### Resizing with nine patch

![Nine Patch ](https://raw.githubusercontent.com/koreezgames/phaser3-ninepatch-plugin/master/assets/np.gif)

#### Resizing without nine patch

![Not Nine Patch](https://raw.githubusercontent.com/koreezgames/phaser3-ninepatch-plugin/master/assets/nnp.gif)

Key features:

-   Blazing fast
-   Low memory usage
-   Easy to use API

## What is Nine-Patch?

A **Nine-Patch** is an image format that uses extra information, defining what parts of it should be scaled (and which are not) when the image is rendered in a larger or smaller size. This technique is very useful for creating UI components like buttons, panels, containers, etc.

Using this technique, you can define the background of multiples UI components like panels or buttons with the same image asset. You can also create large panels with a reduced image asset so is very useful for optimizing your project resources.

![Phaser3 Nine Patch](https://raw.githubusercontent.com/koreezgames/phaser3-ninepatch-plugin/master/assets/what-is-nine-patch.png)

## Getting Started

First you want to get a fresh copy of the plugin. You can get it from this repo or from npm, ain't that handy.

[![npm](https://img.shields.io/npm/dt/@koreez/phaser3-ninepatch.svg)](https://www.npmjs.com/package/@koreez/phaser3-ninepatch)

```
npm install @koreez/phaser3-ninepatch --save
```

## Usage

### Load the plugin

You still need to load the plugin in your game. This is done just like any other global plugin in Phaser3.
So, to load the plugin, include it in phaser game config.

```javascript
{
  type: Phaser.WEBGL,
  width: 540,
  height: 960,
  backgroundColor: '#626262',
  parent: 'gameContainer',
  scene: [GameScene],
  plugins: {
    global: [{ key: 'NinePatchPlugin', plugin: NinePatchPlugin, start: true }],
  },
}
```

The plugin will patch your Phaser game with additional add/make methods so the NinePatch game object fits up in Phaser like any normal object!

### Adding a container

Also adding and creating of 9-patch images is exposed through Phaser's GameObjectCreator and GameObjectFactory.
So you can add a 9-patch image to your game just like any other image once it's preloaded!

```javascript
//We specify the x and y position, the width and height and the key for the image of the image. It will be automaticly scaled!
this.add.ninePatch(100, 100, 200, 400, "my-image", null, {
    top: 10, // Amount of pixels for top
    bottom: 20, // Amount of pixels for bottom
    left: 30, // Amount of pixels for left
    right: 40 // Amount of pixels for right
});
```

Or if you'd want to do something with it first:

```javascript
//Two options here, first we use Phaser
var ninePatch = this.make.ninePatch({
    key: "my-image",
    width: 200,
    height: 200,
    patchesConfig: {
        top: 10, // Amount of pixels for top
        bottom: 20, // Amount of pixels for bottom
        left: 30, // Amount of pixels for left
        right: 40 // Amount of pixels for right
    }
});
ninePatch.x = 20;
ninePatch.y = 20;
this.add.existing(ninePatch);

//Or we use the Constructor
var ninePatch = new NinePatch(this, 0, 0, 200, 50, "my-image", null, {
    top: 10, // Amount of pixels for top
    bottom: 20, // Amount of pixels for bottom
    left: 30, // Amount of pixels for left
    right: 40 // Amount of pixels for right
});
ninePatch.x = 50;
ninePatch.y = 50;
this.add.existing(nineSlice);
```

### Using a Texture Atlas

It's also possible to use an image that's located inside a Texture atlas. The only difference then is that you don't have to preload the image, instead you use the object's constructor and pass the framedata directly on creation:

```javascript
//Add an nineslice image with an texture atlas
var ninePatch = new NinePatch(
    this, // Phaser.Scene
    150, // x position
    100, // y position
    200, // expected width
    100, // expected height
    "buttons", // atlas key
    "btn_clean.png", // Image frame
    {
        top: 20, // Amount of pixels for top
        bottom: 23, // Amount of pixels for bottom
        left: 27, // Amount of pixels for left
        right: 28 // Amount of pixels for right
    }
);
this.add.existing(ninePatch);
```

### Resize method

9-patch image has a resize method which lets you expand/shorten the width or length of the image.
When using resize method, make sure values are not lower than the width of the image corners

```javascript
var ninePatch = this.add.ninePatch(5, 5, 48, 48, "image", null, {
    top: 5 // Amount of pixels for top, bottom, left, right
});
ninePatch.resize(100, 200);
```

## Credits

Big thanks to this great repo:

https://github.com/orange-games/phaser-nineslice

## License

[MIT](LICENSE)


CHANGELOG.md
# Phaser3 Nine Patch Changelog:

## Phaser3 Nine Patch 2.0.0

### v2.0.1

#### Added

-   Add `setTint`, `setTintFill` and `clearTint` methods to be able to change tint.

### v2.0.0

#### Added

-   Now NinePatch class extends Phaser.GameObjects.Container instead of Phaser.GameObjects.RenderTexture to not add additional draws and calls.

### v1.1.5

#### Fixed

-   Minor fixings.

### v1.1.4

#### Fixed

-   Minor fixings.

### v1.1.3

#### Fixed

-   Fix canvas rendering issue.

### v1.1.2

#### Added

-   Add `setTexture` and `setFrame` methods to be able to change texture and/or frame dynamically (see #18).

#### Fixed

-   Fix `addToScene` to be optional for nine patch creator.

### v1.1.1 - Planned stable version

-   fixed resize method to be chainable


example/index.ts
/// <reference path="../types/phaser.d.ts" />

import "phaser";
import { NinePatch } from "../src/com/koreez/plugin/ninepatch/NinePatch";
import { NinePatchPlugin } from "../src/com/koreez/plugin/NinePatchPlugin";

export default class Game extends Phaser.Game {
    public centerX: number;
    public centerY: number;
    constructor(config: GameConfig) {
        super(config);
    }
}

function preload() {
    console.log("preload");
    this.load.image("metalPanel_blueCorner", "static/assets/metalPanel_blueCorner.png");
    this.load.image("squareGreen", "static/assets/squareGreen.png");
    this.load.image("squareYellow", "static/assets/squareYellow.png");
}

function create() {
    console.log("create");
    // We specify the x and y position, the width and height and the key for the image of the image.
    // It will be automatically scaled!
    const metalPanel = this.add.ninePatch(200, 200, 300, 300, "metalPanel_blueCorner", null, {
        bottom: 11, // Amount of pixels for bottom
        left: 11, // Amount of pixels for left
        right: 53, // Amount of pixels for right
        top: 32 // Amount of pixels for top
    });
    let metalPanelDir = 0.6;
    metalPanel.setTint(0xff0000);
    this.tweens.add({
        /* tslint:disable */
        targets: metalPanel,
        duration: 3000,
        y: 190,
        onUpdate: (tween: any) => {
            metalPanel.resize(metalPanel.width + metalPanelDir, metalPanel.height + metalPanelDir);
        },
        onYoyo: (tween: any) => {
            metalPanelDir = -1 * metalPanelDir;
        },
        onRepeat: (tween: any) => {
            metalPanelDir = -1 * metalPanelDir;
        },
        repeat: -1,
        yoyo: true
        /* tslint:enable */
    });
    setTimeout(() => {
        metalPanel.setTintFill(0xff0000);
        setTimeout(() => {
            metalPanel.clearTint(0xff0000);
        }, 3000);
    }, 3000);
    let squareGreen = this.make.ninePatch({
        /* tslint:disable */
        key: "squareGreen",
        width: 100,
        height: 50,
        patchesConfig: {
            bottom: 14, // Amount of pixels for bottom
            left: 6, // Amount of pixels for left
            right: 6, // Amount of pixels for right
            top: 10 // Amount of pixels for top
        }
        /* tslint:enable */
    });

    squareGreen.x = 450;
    squareGreen.y = 50;
    this.add.existing(squareGreen);
    let squareGreenDir = 1;
    this.tweens.add({
        /* tslint:disable */
        targets: squareGreen,
        duration: 3000,
        x: 550,
        onUpdate: (tween: any) => {
            squareGreen.resize(squareGreen.width + squareGreenDir, squareGreen.height);
        },
        onYoyo: (tween: any) => {
            squareGreenDir = -1 * squareGreenDir;
        },
        onRepeat: (tween: any) => {
            squareGreenDir = -1 * squareGreenDir;
        },
        repeat: -1,
        yoyo: true
        /* tslint:enable */
    });

    this.cache.custom.ninePatch.add("squareYellow", {
        /* tslint:disable */
        bottom: 14, // Amount of pixels for bottom
        left: 6, // Amount of pixels for left
        right: 6, // Amount of pixels for right
        top: 10 // Amount of pixels for top
        /* tslint:enable */
    });

    const squareYellow = this.add.ninePatch(550, 500, 300, 100, "squareYellow");
    let squareYellowDir = 0.6;
    this.tweens.add({
        /* tslint:disable */
        targets: squareYellow,
        duration: 3000,
        x: 150,
        repeat: -1,
        yoyo: true,
        onUpdate: (tween: any) => {
            squareYellow.resize(squareYellow.width + squareYellowDir, squareYellow.height);
        },
        onYoyo: (tween: any) => {
            squareYellowDir = -1 * squareYellowDir;
        },
        onRepeat: (tween: any) => {
            squareYellowDir = -1 * squareYellowDir;
        }
        /* tslint:enable */
    });

    // Or we use the Constructor
    const newNinePatch = new NinePatch(this, 630, 300, 200, 50, "squareGreen", null, {
        /* tslint:disable */
        bottom: 14, // Amount of pixels for bottom
        left: 6, // Amount of pixels for left
        right: 6, // Amount of pixels for right
        top: 10 // Amount of pixels for top
        /* tslint:enable */
    });
    this.add.existing(newNinePatch);
    let newNinePatchDir = 0.6;
    this.tweens.add({
        /* tslint:disable */
        targets: newNinePatch,
        duration: 3000,
        rotation: Math.PI,
        repeat: -1,
        yoyo: true,
        onUpdate: (tween: any) => {
            newNinePatch.resize(newNinePatch.width + newNinePatchDir, newNinePatch.height);
        },
        onYoyo: (tween: any) => {
            newNinePatchDir = -1 * newNinePatchDir;
        },
        onRepeat: (tween: any) => {
            newNinePatchDir = -1 * newNinePatchDir;
        }
        /* tslint:enable */
    });
    setTimeout(() => {
        newNinePatch.setTexture("squareYellow");
    }, 4000);
}

document.onreadystatechange = () => {
    if (document.readyState === "complete") {
        const config = {
            banner: true,
            height: 600,
            plugins: {
                global: [{ key: "NinePatchPlugin", plugin: NinePatchPlugin, start: true }]
            },
            scene: {
                create,
                preload
            },
            type: Phaser.AUTO,
            width: 800
        };
        (window as any).game = new Game(config);
    }
};
